File: tests/conformance/dump-sem.ts
/**
 * TS side of the *semantics* conformance dump (contracts.md §5). For each
 * fixture it loads the stock vocab, builds the project symbol table, resolves
 * every reference and runs the portable validator subset, then emits a
 * normalised `{ diagnostics, resolved }` object that must be byte-identical to
 * the Kotlin `SemanticsConformanceDump` output.
 *
 * Normalisation (both runtimes):
 *  - `symbols`: sorted full qnames of every definition in the scenario's own
 *    documents (stock vocab excluded). This is what compares the schema-code
 *    prefix directly, so kind-derived defaults (pkg-schema-defaults) are checked
 *    even when a fixture has no resolvable references.
 *  - `resolved`: sorted `"<refPath> => <resolvedQname>"` strings (one per
 *    reference that resolves) — no source positions.
 *  - `diagnostics`: sorted diagnostic-code strings (code only; severity is
 *    consumer policy and positions are impl-specific).
 */
import type { Document } from '@modeler/parser';
import {
  ProjectSymbolTable,
  Resolver,
  PackageGraphBuilder,
  collectAllReferences,
  enclosingQnameOf,
  resolveManifest,
} from '@modeler/semantics';
import { lintDocument, lintProject, recommendedConfig, RULES } from '@modeler/lint';

// The portable validator subset (matching the Kotlin dump): structural + search +
// reference + import checks per document, plus project-level duplicate
// definition/mapping. Excludes impl-specific rules (file-ordering, package
// declarations, graph, circular dependency).
const PORTABLE_RULE_IDS = new Set([
  'entity-no-attributes',
  'table-no-columns',
  'column-missing-type',
  'attribute-missing-type',
  'missing-description',
  'entity-attribute-not-found',
  'primary-key-column-not-found',
  'fuzzy-without-searchable',
  'duplicate-search-property',
  'unresolved-reference',
  'ambiguous-reference',
  'unimported-reference',
  'unused-import',
  'duplicate-import',
  'wildcard-with-no-matches',
  'duplicate-definition',
  'duplicate-mapping',
]);
const PORTABLE_RULES = [...RULES.values()].filter((r) => PORTABLE_RULE_IDS.has(r.id));

export interface SemDump {
  resolved: string[];
  diagnostics: string[];
  symbols: string[];
}

/** One parsed document in a (possibly multi-file) scenario. */
export interface SemDocInput {
  ast: Document;
  uri: string;
}

/** Single-document dump — the common case. A 1-element delegation to {@link dumpSemDocs}. */
export function dumpSem(ast: Document | undefined, uri: string, stock: Map<string, Document>): SemDump {
  if (!ast) return dumpSemDocs([], stock);
  return dumpSemDocs([{ ast, uri }], stock);
}

/**
 * Multi-document dump: builds ONE project symbol table from the stock vocab plus
 * every document in the scenario, then resolves references and runs the portable
 * validator subset across all of them. This is what exercises cross-file
 * resolution (same-package siblings, named/wildcard imports) — see the
 * `fixtures/<scenario>/` directories. For a single document it is byte-identical
 * to the previous single-doc dump.
 */
export function dumpSemDocs(docs: SemDocInput[], stock: Map<string, Document>): SemDump {
  const symbols = new ProjectSymbolTable();
  for (const [name, doc] of stock) {
    symbols.upsertDocument(`stock://${name}.ttrm`, doc, 'cnc', 'role', '');
  }

  // Upsert every document FIRST so cross-document lookups (getByPackage,
  // named/wildcard imports, getBySuffix) see the whole project.
  const metas = docs.map(({ ast, uri }) => {
    // '' (no directive) ⇒ the semantics layer derives the schema per definition
    // from its kind. Must match the Kotlin dump (`?: ""`) for byte-identical output.
    const schemaCode = ast.schemaDirective?.schemaCode ?? '';
    const namespace = ast.schemaDirective?.namespace ?? '';
    const packageName = ast.packageDecl?.name ?? '';
    symbols.upsertDocument(uri, ast, schemaCode, namespace, packageName);
    return { ast, uri, schemaCode, namespace, packageName };
  });

  const resolver = new Resolver(symbols);
  const deps = { manifest: resolveManifest(undefined, ''), symbols, resolver };
  const config = recommendedConfig();
  const projectDocs = new Map(metas.map((m) => [m.uri, m.ast]));
  const packageGraph = new PackageGraphBuilder(symbols, projectDocs).build();

  const resolved: string[] = [];
  const diagnostics: string[] = [];
  for (const m of metas) {
    for (const { ref, ownerDef } of collectAllReferences(m.ast)) {
      const enclosingQname = enclosingQnameOf(ownerDef, m.schemaCode, m.namespace, m.packageName);
      const res = resolver.resolveReference(
        { path: ref.path, parts: ref.parts },
        { schemaCode: m.schemaCode, namespace: m.namespace, enclosingQname, imports: m.ast.imports, packageName: m.packageName }
      );
      if (res.resolved) resolved.push(`${ref.path} => ${res.symbol.qname}`);
    }
    diagnostics.push(...lintDocument(m.uri, m.ast, deps, config, PORTABLE_RULES).map((d) => d.code));
  }
  // Project-scoped rules (duplicate definition/mapping) run once across all docs.
  for (const diags of lintProject(projectDocs, packageGraph, deps, config, PORTABLE_RULES).values()) {
    diagnostics.push(...diags.map((d) => d.code));
  }

  // Full qnames of the scenario's own definitions (stock vocab excluded).
  const symbolQnames = symbols
    .all()
    .filter((e) => !e.documentUri.startsWith('stock://'))
    .map((e) => e.qname);

  resolved.sort();
  diagnostics.sort();
  symbolQnames.sort();
  return { resolved, diagnostics, symbols: symbolQnames };
}

/** Keys ordered `diagnostics`, `resolved`, `symbols`; matches the Kotlin render. */
export function renderSem(d: SemDump): string {
  return JSON.stringify({ diagnostics: d.diagnostics, resolved: d.resolved, symbols: d.symbols }, null, 4);
}

File: tests/conformance/dump_sem.py
"""Python-side SEMANTICS conformance dump (contracts §5.1).

Mirror of `dump-sem.ts` / `run-ts-sem.ts`. For each fixture — a single `.ttr`
file or a multi-document subdirectory — it loads the stock cnc vocab, builds ONE
`SymbolTable`, resolves every reference, runs the portable validator subset, and
emits a normalised `{ diagnostics, resolved, symbols }` object that must be
byte-identical to the committed TS golden `out-ts-sem/`.

Normalisation (identical to `dump-sem.ts`):
  - `resolved`  — sorted `"<refPath> => <resolvedQname>"`, resolved with the
                  document's RAW schema code (no per-kind default — this is what
                  exercises the schema-prefix in the qname), one per ref.
  - `diagnostics` — sorted `DiagnosticCode.value` strings from the portable
                    validator (severity/positions not compared).
  - `symbols`   — sorted full qnames of the scenario's own defs (stock excluded).

JSON formatting matches `JSON.stringify(value, null, 4)` + trailing newline.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

from ttr_parser import parse_file, parse_string
from ttr_parser.model import ParseResult
from ttr_parser.semantics import (
    ResolutionContext,
    Resolved,
    Resolver,
    SymbolTable,
    Validator,
)
from ttr_parser.semantics.references import collect_all_references, enclosing_qname_of

REPO_ROOT = Path(__file__).resolve().parent.parent.parent
FIXTURES = REPO_ROOT / "tests" / "conformance" / "fixtures"
OUT_PY_SEM = REPO_ROOT / "tests" / "conformance" / "out-py-sem"
OUT_TS_SEM = REPO_ROOT / "tests" / "conformance" / "out-ts-sem"
# The canonical stock vocab (the same content dump-sem.ts loads via
# @modeler/semantics); the build copies this into the package for runtime use.
STOCK_FILE = REPO_ROOT / "packages" / "semantics" / "src" / "stock" / "cnc-roles.ttr"
STOCK_URI = "stock://cnc-roles.ttr"


def dump_sem_docs(docs: list[tuple[ParseResult, str]]) -> str:
    """Build one symbol table from stock + every doc, resolve, validate, render."""
    symbols = SymbolTable()

    # Stock first, so cnc.* auto-imports resolve to the doubled cnc.cnc.role.*.
    stock_result = parse_string(STOCK_FILE.read_text(encoding="utf-8"), STOCK_URI)
    symbols.upsert_document(STOCK_URI, stock_result)

    # Upsert every document FIRST so cross-document lookups see the whole project.
    metas: list[tuple[ParseResult, str, str, str]] = []
    for result, uri in docs:
        directive = result.schema_directive
        schema_code = directive.schema_code if directive else ""
        namespace = (directive.namespace if directive else "") or ""
        package = result.package_name or ""
        symbols.upsert_document(uri, result, package_name=package)
        metas.append((result, schema_code, namespace, package))

    resolver = Resolver(symbols)

    resolved: list[str] = []
    for result, schema_code, namespace, package in metas:
        for collected in collect_all_references(result.definitions):
            enclosing = enclosing_qname_of(
                collected.owner_def, schema_code, namespace, package
            )
            res = resolver.resolve_reference(
                collected.path,
                ResolutionContext(
                    schema_code=schema_code,
                    namespace=namespace,
                    imports=result.imports,
                    package_name=package,
                    enclosing_qname=enclosing,
                ),
            )
            if isinstance(res, Resolved):
                resolved.append(f"{collected.path} => {res.symbol.qname.value}")

    diagnostics = [
        d.code.value
        for d in Validator().validate([r for r, _ in docs], symbols)
    ]

    symbol_qnames = [
        e.qname.value
        for e in symbols.get_all()
        if not e.source_file.startswith("stock://")
    ]

    resolved.sort()
    diagnostics.sort()
    symbol_qnames.sort()
    payload = {
        "diagnostics": diagnostics,
        "resolved": resolved,
        "symbols": symbol_qnames,
    }
    return json.dumps(payload, ensure_ascii=False, indent=4) + "\n"


def single_fixtures() -> list[Path]:
    return sorted(p for p in FIXTURES.iterdir() if p.is_file() and p.suffix == ".ttr")


def dir_fixtures() -> list[Path]:
    return sorted(p for p in FIXTURES.iterdir() if p.is_dir())


def main() -> int:
    OUT_PY_SEM.mkdir(parents=True, exist_ok=True)
    status = 0

    for fixture in single_fixtures():
        result = parse_file(fixture)
        if result.errors:
            print(f"warning: {fixture.name}: parse errors", file=sys.stderr)
        text = dump_sem_docs([(result, fixture.name)])
        (OUT_PY_SEM / (fixture.stem + ".json")).write_text(text, encoding="utf-8")

    for directory in dir_fixtures():
        docs: list[tuple[ParseResult, str]] = []
        for sub in sorted(directory.iterdir()):
            if sub.is_file() and sub.suffix == ".ttr":
                result = parse_file(sub)
                if result.errors:
                    print(f"warning: {directory.name}/{sub.name}: parse errors", file=sys.stderr)
                docs.append((result, f"{directory.name}/{sub.name}"))
        text = dump_sem_docs(docs)
        (OUT_PY_SEM / (directory.name + ".json")).write_text(text, encoding="utf-8")

    print(
        f"dumped {len(single_fixtures())} single-file + {len(dir_fixtures())} "
        f"multi-doc semantics fixtures to {OUT_PY_SEM}"
    )
    return status


if __name__ == "__main__":
    raise SystemExit(main())
